app/models/journable/historic_active_record_relation.rb

```diff
@@ -65,145 +58,108 @@ def initialize(relation, timestamp:)
   #
-  def build_arel(connection, aliases = nil)
-    relation = self
+  # The way this is done is by prepending a CTE to the list of possible already defined CTEs
+  # under the name of the model the historic data is queried on. The CTE will then be on the
+  # join of:
+  # * the model's data journal (for the property values at the time)
+  # * the journals table itself (for the update timestamp)
+  # * the model table (for the creation timestamp)
+  # The result is that all later statements, including later CTEs, will pick up data from that
+  # first CTE. This makes the fact that data is fetched not from the model but from the journals
+  # transparent so most queries that work on the model table just continue to work on the journals
+  # join. The only currently known exceptions to this are conditions on custom fields. That is
+  # because those should also work on the historic data and not the current one, so the statement
+  # needs to be rewritten.
+  #
+  # A statement on work packages might then look like this:
+  #
+  # WITH work_packages AS (
+  #     [SQL join on work_package_journals, journals and work_packages]
+  #   ),
+  #   other_cte AS ([...with a FROM work_packages...])
+  #
+  # SELECT * from work_packages
 
-    relation = switch_to_journals_database_table(relation)
-    relation = substitute_join_tables_in_where_clause(relation)
-    relation = substitute_database_table_in_where_clause(relation)
-    relation = add_timestamp_condition(relation)
-    relation = add_join_on_journables_table_with_created_at_column(relation)
-    relation = add_join_projects_on_work_package_journals(relation)
-    relation = select_columns_from_the_appropriate_tables(relation)
+  def build_arel(connection, aliases = nil)
+    substitute_join_tables_in_where_clause(self)
 
-    # Based on the previous modifications, build the algebra object.
-    arel = relation.call_original_build_arel(connection, aliases)
-    arel = modify_where_clauses(arel)
-    arel = modify_order_clauses(arel)
-    modify_joins(arel)
+    # Based on the previous modifications, build the algebra object and prepend
+    # the journals CTE.
+    add_historic_model_cte(super)
   end
 
-  def call_original_build_arel(connection, aliases = nil)
-    original_build_arel(connection, aliases)
-  end
+  private
+
+  def add_historic_model_cte(arel)
+    historic_models_cte = Arel::Nodes::As.new(Arel::Table.new(model.table_name),
+                                              historic_models_statement.arel)
 
-  def eager_loading?
-    false
+    if arel.ast.with
+      arel.ast.with.expr.unshift(historic_models_cte)
+    else
+      arel.ast.with = Arel::Nodes::With.new([historic_models_cte])
+    end
+
+    arel
   end
 
-  private
+  def historic_models_statement
+    relation = Journal
+                 .joins(historic_models_data_journals_join)
+                 .joins(historic_models_original_models_join)
+                 .select(historic_models_selects)
```

Comment:
It is possible to at least partially join using dsl (`model.joins(:journals)`, `model.journal_class.joins(:journal)`), but I failed to find a way to combine those